realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -146,6 +151,23 @@ public boolean handleMessage(Message message) {
         return true;
     }
 
+    /**
+     * Properly handles when an async transaction completes. This will be treated as a REALM_CHANGED event when
+     * determining which queries to re-run and when to notify listeners.
+     * <p>
+     * NOTE: This is needed as it is not possible to combine a `Message.what` value and a callback runnable. So instead
+     * of posting two messages, we post a runnable that runs this method. This means it is possible to interpret
+     * `REALM_CHANGED + Runnable` as one atomic message.
+     *
+     * @param onSuccess onSuccess callback to run for the async transaction that completed.
+     */
+    public void handleAsyncTransactionCompleted(Runnable onSuccess) {
+        if (onSuccess != null) {
+            pendingOnSuccessAsyncTransactionCallbacks.add(onSuccess);
+        }
+        realmChanged(false);
```

Comment:
What is the reason that not using some thread safe container for callbacks and still trigger changes by `REALM_CHANGED`? 

The good thing about `REALM_CHANGED` is we will have 2 events in the queue for the worst case. But posting a callback mean below code will be a nightmare:

```
loop( 500 times) {
    executeAsyncTransaction(...)
}
```